routes/ads.js

```diff
@@ -28,31 +28,99 @@ router.get("/", async function (req, res, next) {
                 status: "success",
             });
         }
-        
-        
+
         const query = `SELECT id, cityId, image, link, createdAt FROM ${tables.ADVERTISEMENTS} WHERE (cityId IS NULL OR cityId = ?) AND enabled = True`
         const response = await database.callQuery(query, [Number(filter.cityId)])
         const data = response.rows
 
         const dataReturn = data[(Math.floor(Math.random() * data.length))]
 
-        if(dataReturn) {
-            await database.update(tables.ADVERTISEMENTS, { lastShown: currentDate}, { id: dataReturn.id })
+        if (dataReturn) {
+            await database.update(tables.ADVERTISEMENTS, { lastShown: currentDate }, { id: dataReturn.id })
         }
         res.status(200).json({
             status: "success",
             data: dataReturn,
         });
-    } catch(error) {
+    } catch (error) {
+        return next(new AppError(error));
+    }
+});
+
+router.get("/list", async function (req, res, next) {
+    try {
+        /*
+        Query Params:
+        - cityId: number 
+        - skipAdIds: string (comma separated adIds)
+        - count: number (default 5)
+        - sort: string (default lastShown)
+        - sortDesc: boolean (default false)
+        */
+        const skipAdIdsStr = req.query.skipAdIds;
+        const returnAdsCount = req.query.count ? Number(req.query.count) : 5;
+        const sort = req.query.sort ? req.query.sort : "lastShown";
+        const sortDesc = req.query.sortDesc === "true" ? true : false;
+        let cityId = req.query.cityId ? Number(req.query.cityId) : null;
+
+        const queryParams = [];
+        if (!cityId || isNaN(cityId)) {
+            cityId = null;
+        } else {
+            queryParams.push(cityId);
+        }
+
+        let skipAdIds = skipAdIdsStr ? skipAdIdsStr.split(",").map(Number) : [];
+        skipAdIds = skipAdIds.filter((id) => !isNaN(id));
+
+        const currentDate = new Date();
+
+        let query = `SELECT id, cityId, image, link, createdAt FROM ${tables.ADVERTISEMENTS} 
+        WHERE 
+            ${cityId ? "cityId = ?" : "1=1"}
+        AND enabled = True`
+
+        let countQuery = `SELECT COUNT(*) as count FROM ${tables.ADVERTISEMENTS}
+        WHERE 
+            ${cityId ? "cityId = ?" : "1=1"}
+            AND enabled = True
+        `
+        if (skipAdIds.length > 0) {
+            query += ` AND id NOT IN (${skipAdIds.map(() => '?').join(",")})`
+            countQuery += ` AND id NOT IN (${skipAdIds.map(() => '?').join(",")})`
+            queryParams.push(...skipAdIds);
+        }
+
+        query += ` ORDER BY ${sort} ${sortDesc ? "DESC" : "ASC"} LIMIT ?`
+        queryParams.push(returnAdsCount);
+
+        const response = await database.callQuery(query, queryParams)
+        const data = response.rows;
+
+        const countResponse = await database.callQuery(countQuery, queryParams)
+        const count = countResponse.rows[0].count;
+
+        Promise.all(
```

Comment:
I would suggest doing this in 1 query itself. It will create two many connections to the DB for every call and is very inefficient. Do something like database.update(tables.ADVERTISEMENTS, { lastShown: currentDate }, { id: [listOfIds] })

Reply:
Fixed